src/coreclr/jit/optimizer.cpp

```diff
@@ -4594,15 +4594,363 @@ PhaseStatus Compiler::optUnrollLoops()
+
+        thenBlock = thenBlockNext;
+    }
+
+    // Nothing found. Still valid to continue.
+    return true;
+}
+
+//-----------------------------------------------------------------------------
+// IfConvertFindFlow
+//
+// Find a valid if conversion flow from m_startBlock to a final block.
+// There might be multiple Then and Else blocks in the flow - use m_checkLimit to limit this.
+//
+// Notes:
+//   Sets m_flowFound, m_finalBlock, m_doElseConversion and m_mainOper.
+//
+void OptIfConversionDsc::IfConvertFindFlow()
+{
+    // First check for flow with no else case. The final block is the destination of the jump.
+    m_doElseConversion = false;
+    m_finalBlock       = m_startBlock->bbJumpDest;
+    assert(m_finalBlock != nullptr);
+    if (!IfConvertCheckThenFlow() || m_flowFound)
+    {
+        // Either the flow is invalid, or a flow was found.
+        return;
+    }
+
+    // Look for flows with else blocks. The final block is the block after the else block.
+    m_doElseConversion = true;
+    for (int elseLimit = 0; elseLimit < m_checkLimit; elseLimit++)
+    {
+        BasicBlock* elseBlock = m_finalBlock;
+        if (elseBlock == nullptr || !IfConvertCheckInnerBlockFlow(elseBlock))
+        {
+            // Need a valid else block in a valid flow .
+            return;
+        }
+
+        m_finalBlock = elseBlock->GetUniqueSucc();
+
+        if (!IfConvertCheckThenFlow() || m_flowFound)
+        {
+            // Either the flow is invalid, or a flow was found.
+            return;
+        }
+    }
+}
+
+//-----------------------------------------------------------------------------
+// IfConvertCheckStmts
+//
+// From the given block to the final block, check all the statements and nodes are
+// valid for an If conversion. Chain of blocks must contain only a single assignment
+// and no other operations.
+//
+// Arguments:
+//   fromBlock  -- Block inside the if statement to start from (Either Then or Else path).
+//   foundOperation -- Returns the found operation.
+//
+// Returns:
+//   If everything is valid, then set foundOperation to the assignment and return true.
+//   Otherwise return false.
+//
+bool OptIfConversionDsc::IfConvertCheckStmts(BasicBlock* fromBlock, IfConvertOperation* foundOperation)
+{
+    bool found = false;
+
+    for (BasicBlock* block = fromBlock; block != m_finalBlock; block = block->GetUniqueSucc())
+    {
+        assert(block != nullptr);
+
+        // Can all the nodes within the block be made to conditionally execute?
+        for (Statement* const stmt : block->Statements())
+        {
+            GenTree* tree = stmt->GetRootNode();
+            switch (tree->gtOper)
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->gtGetOp1();
+                    GenTree* op2 = tree->gtGetOp2();
+
+                    // Only one per operation per block can be conditionally executed.
+                    if (found || op2->OperIs(GT_SELECT))
```

Comment:
I wanted to leave this to a later PR.....

You'd have a single compare node and two select nodes. Both select nodes would somehow have to connect to the same compare. Everything would have to be contained to prevent the compare going to registers.

I think you could do it by allowing the compare node of a select to be another select node.

`SELECT( SELECT(COND, A, B), C, D )`

If we allow that then that allows for things like:
`if (condition) { a=1; b=2; } else { b=5; c=4; d=9; }`

The pass would have to catch there is a assignment to 'b' in both the then and else.

You'd also limit the number of statements - llvm maxes out at 3, I think.

Reply:
Ignore my comment above.

To get to this nested SELECT state, that means we've got two statements:
```
JTRUE(COND2)
ASG(X, SELECT(COND1, A, B))
```
That would have to be combined into the following:
```
ASG(X, SELECT(AND(COND1,COND2), A, B))
```

Optimising that would in my original If conversion patch. It was suggested at the time that it should be separated out so that the optimise bools pass created an `AND` compare chain.

That's definitely for additional patch (I've got a version of it mostly working, except for a few bugs).

Reply:
What I'm after here is not optimization of nested selects. I agree that should be done in a separate PR if there is an opportunity here.

What I'm after is that we remove the `GT_SELECT` check here. I see no reason why it needs to be special cased. What would break?

Reply:
If it was allowed then:
```
JTRUE(COND2)
ASG(X, SELECT(COND1, A, B))
```
would become:
```
ASG(X, SELECT(COND2, SELECT(COND1, A, B), X))
```
.... Which should work. The optimise bools version would be better, but that code doesn't exist yet

Ok, will remove the check.